contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,378 @@
+The device plugin should also run in containers so that kubernetes can
+deploy them and restart the plugins when they fail.
+However, we should not prevent the user from deploying a bare metal device
+plugin.
+
+Deploying the device plugins though DemonSets makes sense as the cluster
+admin would be able to specify which machines it wants the device plugins to
+run on, the process is similar to any kubernetes action and does not require
+to change any parts of Kubernetes.
+
+Additionally, for integrated solutions such as `kubeadm` we can add support
+to auto-deploy vetted Device Plugins.
+Thus not fragmenting once more the Kubernetes ecosystem.
+
+For users installing Kubernetes without using an integrated solution such
+as `kubeadm` they would use the examples that we would provide at:
+`https://github.com/kubernetes/kubernetes/tree/master/examples/device-plugin.yaml`
+
+YAML example:
+```yaml
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+spec:
+    template:
+        metadata:
+            labels:
+                - name: device-plugin
+        spec:
+            containers:
+                name: device-plugin-ctr
+                image: nvidia/device-plugin:1.0
+                volumeMounts:
+                  - mountPath: /device-plugin
+                  - name: device-plugin
+           volumes:
+             - name: device-plugin
+               hostPath:
+                   path: /var/run/kubernetes
+```
+
+## API Changes
+### ResourceType
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called the ResourceType.
+It is defined as follows:
+
+```golang
+type ResourceType struct {
+	Kind string
+	Name string
+	Properties map[string]string
+}
+```
+
+Because the current API (Capacity) can not be extended to support ResourceType,
+we will need to create two new attributes in the NodeStatus structure:
+  * `CapacityV2`: Describing the capacity of the node
+  * `Available`: Describing the available resources
+
+```golang
+type NodeStatus struct {
+	Capacity ResourceList
+	CapacityV2 []ResourceType
+
+	Available[]ResourceType
+}
+```
+
+We also introduce the `Allocated` field in the pod's status so that user
+can know what devices were assigned to the pod. It could also be useful in
+the case of monitoring
+
+```golang
+type PodStatus struct {
+	Allocated[]ResourceType
+}
+```
+
+## Device Plugin
+
+The device plugin is structured in 4 parts:
+1. Registration: The device plugin advertises it's presence to Kubelet
+2. Discovery: Kubelet calls the device plugin to list it's devices
+3. Allocate / Deallocate: When creating/deleting containers requesting the
+   devices advertised by the device plugin, Kubelet calls the device plugin's
+   `Allocate` and `Deallocate` functions.
+4. Cleanup: Kubelet terminates the communication through a "Stop"
```

Comment:
@RenaudWasTaken api name used down in this doc is  `Shutdown`, should the same be used here instead of "Stop". 
@cmluciano Diagram below shows "unload drivers", so i think this will clean up what daemonset installed.